modules/battery_manager/php/battery_manager.class.inc

```diff
@@ -0,0 +1,249 @@
+        try {
+          $db->update('test_battery', array('Active' => $status), array('ID' => $id));
+          return $this->response(200, 'Availability Updated Successfully');
+        } catch (\DatabaseException $e) {
+          return $this->response(400, 'Could not update entry in the Test Battery');
+        }
+    }
+
+    function addTest($db, $user, $data) {
+        if (!$user->hasPermission('battery_manager_edit')) {
+            return $this->response(400, 'Edit Permission Denied');
+        }
+
+        if (!$this->isDuplicate($db, $data)) {
+            return $this->response(400, 'This Test already exists in the database');
+        }
+
+        $test = array(
+                'Test_name'    => $data['testName'],
+                'AgeMinDays'   => $data['ageMinDays'],
+                'AgeMaxDays'   => $data['ageMaxDays'],
+                'Stage'        => $data['stage'],
+                'SubprojectID' => $data['subproject'],
+                'Visit_label'  => $data['visitLabel'],
+                'CenterID'     => $data['centerId'],
+                'firstVisit'   => $data['firstVisit'],
+                'instr_order'  => $data['instrumentOrder'],
+                'Active'       => 'Y',
+        );
+
+        try {
+          $db->insert('test_battery', $test);
+          return $this->response(200, 'Test Submission Successful');
+        } catch (\DatabaseException $e) {
+          return $this->response(400, 'Could not add entry to the Test Battery');
+        }
+    }
+
+    function isDuplicate($db, $data) {
+        // Build SQL query based on values entered by user
+        $query     = "SELECT Test_name as testName,
+                             AgeMinDays as ageMinDays,
+                             AgeMaxDays as ageMaxDays,
+                             Stage as stage,
+                             SubprojectID as subproject,
+                             Visit_label as visitLabel,
+                             CenterID as centerId,
+                             firstVisit,
+                             instr_order as instrumentOrder
+                      FROM test_battery";
+        // Select duplicate entry from Test Battery
+        $entries = $db->pselect($query, array());
+        
+        foreach ($entries as $entry) {
+            if ($data === $entry) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    function response(int $status, string $body) : ResponseInterface
+    {
+        return (new \LORIS\Http\Response())
+            ->withHeader('Content-Type', 'application/json')
+            ->withStatus($status)
+            ->withBody(new \LORIS\Http\StringStream($body));
+    }
+
+    /**
+     * Converts the results of this menu filter to a JSON format
+     *
+     */
+    function toJSON()
+    {
+        $user = \User::singleton();
+        $tests = [];
+        $testIt = $this->getDataProvisioner()->execute($user);
```

Comment:
I think the \LORIS\Data\Table() class should be used as a helper here instead. Something like:

```
return (new \LORIS\Data\Table())
      ->withDataFrom($this->getDataProvisioner)
      ->toJSON($user);
```

That way we have all the places that do this centralized in one place if we ever take the time to convert it into a lazily evaluated stream rather than loading everything into an array to serialize it.

Reply:
I'm having a lot of trouble with an implementation that uses your template. I cannot get the proper format that I need in order to manipulate the data on the front end in a streamlined way.